devito/propagator.py

```diff
@@ -433,24 +433,29 @@ def get_time_stepping(self):
 
         return body
 
-    def time_substitutions(self, sympy_expr):
-        """This method checks through the sympy_expr to replace the time index with a cyclic index
-        but only for variables which are not being saved in the time domain
-
-        :param sympy_expr: The Sympy expression to process
-        :returns: The expression after the substitutions
-        """
+    def _time_substitutions(self, sympy_expr, subs_dict):
```

Comment:
Could we please use SymPy's built-in generator traversals `sympy.preorder_traversal` instead of actually recursing? That should make things even faster.